compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -969,29 +1095,202 @@ mod test {
         let ssa = ssa.loop_invariant_code_motion();
         let expected = "
         brillig(inline) fn main f0 {
-          b0():
-            jmp b1(i32 1)
-          b1(v0: i32):
-            v4 = lt v0, i32 4
-            jmpif v4 then: b3, else: b2
+          b0(v0: u32):
+            v4 = eq v0, u32 5
+            jmp b1(u32 1)
+          b1(v1: u32):
+            v7 = lt v1, u32 4
+            jmpif v7 then: b2, else: b3
           b2():
+            v9 = div u32 10, v1
+            jmp b4(u32 0)
+          b3():
             return
+          b4(v2: u32):
+            v11 = lt v2, u32 4
+            jmpif v11 then: b5, else: b6
+          b5():
+            jmpif v4 then: b7, else: b8
+          b6():
+            v12 = unchecked_add v1, u32 1
+            jmp b1(v12)
+          b7():
+            constrain v9 == u32 6
+            jmp b8()
+          b8():
+            v14 = unchecked_add v2, u32 1
+            jmp b4(v14)
+        }
+        ";
+
+        assert_normalized_ssa_equals(ssa, expected);
+    }
+}
+
+#[cfg(test)]
+mod control_dependence {
+    use crate::ssa::{opt::assert_normalized_ssa_equals, ssa_gen::Ssa};
+
+    #[test]
+    fn do_not_hoist_unsafe_mul_in_control_dependent_block() {
+        let src = "
+        brillig(inline) fn main f0 {
+          b0(v0: u32, v1: u32):
+            v4 = eq v0, u32 5
+            jmp b1(u32 0)
+          b1(v2: u32):
+            v7 = lt v2, u32 4
+            jmpif v7 then: b2, else: b3
+          b2():
+            jmpif v4 then: b4, else: b5
           b3():
-            v6 = div i32 10, v0
-            jmp b4(i32 0)
-          b4(v1: i32):
-            v8 = lt v1, i32 4
-            jmpif v8 then: b6, else: b5
+            return
+          b4():
+            v8 = mul v0, v1
+            constrain v8 == u32 12
+            jmp b5()
           b5():
-            v11 = unchecked_add v0, i32 1
+            v11 = unchecked_add v2, u32 1
             jmp b1(v11)
-          b6():
-            constrain v6 == i32 6
-            v10 = unchecked_add v1, i32 1
-            jmp b4(v10)
         }
         ";
 
+        let ssa = Ssa::from_str(src).unwrap();
+
+        let ssa = ssa.loop_invariant_code_motion();
+        assert_normalized_ssa_equals(ssa, src);
+    }
+
+    #[test]
+    fn hoist_safe_mul_that_is_non_control_dependent() {
+        let src = "
+        brillig(inline) fn main f0 {
```

Comment:
FYI it should be possible to use "regular" identifiers for these tests after https://github.com/noir-lang/noir/pull/7697

Reply:
I switched to regular identifiers